test/fn/edit-contacts.spec.js

```diff
@@ -0,0 +1,154 @@
+
+// specifying directory paths to use
+const editContactsPath = `data/edit-contacts/documents`;
+const csvDir = `${editContactsPath}/csv`;
+const saveDocsDir = `${editContactsPath}/json_docs`;
+const expectedDocsDir = `${editContactsPath}/expected-json_docs`;
+const filesToUpload = fs.recurseFiles(`${editContactsPath}/server-contact_docs`).filter(name => name.endsWith('.json'));
+const saveJsonDoc = doc => fs.write(`${saveDocsDir}/${doc._id}.doc.json`, toDocs.toSafeJson(doc) + '\n');
+const countFilesInDir = path => fs.fs.readdirSync(path).length;
+const csvFiles = ['contact.csv'];
+
+
+const docs = filesToUpload
+  .map(file => {
+    const doc = fs.readJson(file);
+    return doc;
+  });
+
+const uploadDocuments = (docs) => {
+  return pouch.bulkDocs(docs);
+};
+
+const fetchDocuments = () => {
+  return pouch.allDocs({include_docs: true});
+};
+
+const processDocuments = (docType, csv, contactDocs, args) => {
+  return processDocs(docType, csv, contactDocs, args);
+};
+
+function compareDocuments(){
+  fs.recurseFiles(expectedDocsDir)
+      .map(file => fs.path.basename(file))
+      .forEach(file => {
+        const expected  = fs.readJson(`${expectedDocsDir}/${file}`);
+        const generated = fs.readJson(`${saveDocsDir}/${file}`);
+        delete generated._rev;
+        expect(generated).to.deep.eq(expected);
+      });
+}
+
+describe('edit-contacts', function() {
+
+  before(async () => {
+    await uploadDocuments(docs);
+    sinon.stub(environment, 'pathToProject').get(() => editContactsPath);
+    const pouchDb = sinon.stub();
+    pouchDb.returns(pouch);
+    editContactsModule.__set__('pouch', pouchDb);
+  });
+
+  after(async () => pouch.destroy());
+
+  it(`should do a top-down test well`, async function(){
+
+    await editContactsModule.execute();    
+    assert.equal(countFilesInDir(saveDocsDir),
+                countFilesInDir(expectedDocsDir),
+                `Different number of files in ${saveDocsDir} and ${expectedDocsDir}.`);
+    compareDocuments();
+  }); 
+  
+  it(`should fail when wrong column names are provided`, async function(){
+
+    const parseResult = {
+      colNames: ['enmch'],
+      csvFiles: ['contact.csv'],
+      docDirectoryPath: 'json_docs',
+      force: false,
+    };
+    const extraArgs = sinon.stub();
+    extraArgs.returns(parseResult);
+    editContactsModule.__set__('parseExtraArgs', extraArgs);
+    
+    try {
+      await editContactsModule.execute();
+      assert.fail('should throw an error when wrong column names are provided');
+    } catch (err) {
+      expect(err.message).to.be.equal('The column name(s) specified do not exist.');
+    }
+  }); 
+
+  it(`should fail when protected column names are provided`, async function(){
+
+    const docs = await fetchDocuments();
+    const processCsv = editContactsModule.__get__('processCsv');
+    
+    try {
+      processCsv('contact',['parent'],['0ebca32d-c1b7-5522-94a3-97dd8b3df146','parent_id'],0, [],docs);
```

Comment:
I think this test - and all the following tests - should call `editContactsModule.execute` instead of a rewired function call. 
It's fine to have fine testing of subcomponents, but you should prioritize testing the module as a whole. 
This test doesn't answer the question of: what happens when I pass a protected column name to the module? Does it fail? I don't know. `processCsv` fails, but is it caught? does execution still happen? are there any files created?

Reply:
This totally makes sense, we've now implemented the tests to call `editContactsModule.execute` instead of a rewired function call so that we can test the module as a whole.